File: pluto/src/vis/gl/compiler.ts
// Copyright 2023 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { XY } from "@synnaxlabs/x";

import { RGBATuple } from "@/color";
import { errorCompile, ERROR_BAD_SHADER, ERROR_NOT_COMPILED } from "@/vis/gl/errors";

export interface Compiler {
  /** Compiles and links the program to the given context.  */
  compile: (gl: WebGLRenderingContext) => void;
}

/** A compiler that compiles a static shader program. */
export class StaticCompiler implements Compiler {
  private readonly vertShader: string;
  private readonly fragShader: string;
  private _program: WebGLProgram | null;
  compiled: boolean;

  constructor(vertShader: string, fragShader: string) {
    this._program = null;
    this.vertShader = vertShader;
    this.fragShader = fragShader;
    this.compiled = false;
  }

  get program(): WebGLProgram {
    this.validateCompiled();
    return this._program as WebGLProgram;
  }

  compile(gl: WebGLRenderingContext): void {
    this._program = gl.createProgram();
    this.compiled = true;
    this.compileShader(gl, this.vertShader, gl.VERTEX_SHADER);
    this.compileShader(gl, this.fragShader, gl.FRAGMENT_SHADER);
    gl.linkProgram(this._program as WebGLProgram);
  }

  compileShader(gl: WebGLRenderingContext, shader: string, type: number): void {
    const vs = gl.createShader(type);
    if (vs == null) throw ERROR_BAD_SHADER;
    gl.shaderSource(vs, shader);
    gl.compileShader(vs);
    const compiled = gl.getShaderParameter(vs, gl.COMPILE_STATUS) as GLboolean;
    if (!compiled) {
      const error = gl.getShaderInfoLog(vs);
      gl.deleteShader(vs);
      throw errorCompile(error ?? "unknown");
    }
    gl.attachShader(this.program, vs);
  }

  validateCompiled(): void {
    if (!this.compiled) throw ERROR_NOT_COMPILED;
  }

  use(gl: WebGLRenderingContext): void {
    gl.useProgram(this.program);
  }

  uniformXY(gl: WebGLRenderingContext, name: string, value: XY): void {
    gl.uniform2fv(this.getUniformLoc(gl, name), [value.x, value.y]);
  }

  uniformColor(gl: WebGLRenderingContext, name: string, value: RGBATuple): void {
    gl.uniform4fv(this.getUniformLoc(gl, name), value);
  }

  private getUniformLoc(gl: WebGLRenderingContext, name: string): WebGLUniformLocation {
    const loc = gl.getUniformLocation(this.program, name);
    if (loc == null) throw new Error(`unexpected missing uniform ${name}`);
    return loc;
  }
}

File: slate/compiler/compiler.go
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package compiler

import (
	"github.com/synnaxlabs/slate/analyzer"
	"github.com/synnaxlabs/slate/compiler/core"
	"github.com/synnaxlabs/slate/compiler/statement"
	"github.com/synnaxlabs/slate/compiler/wasm"
	"github.com/synnaxlabs/slate/parser"
	"github.com/synnaxlabs/slate/symbol"
	"github.com/synnaxlabs/slate/types"
	"github.com/synnaxlabs/x/errors"
)

type Config struct {
	Program            parser.IProgramContext
	Analysis           *analyzer.Result
	DisableHostImports bool
}

func paramsToWasm(params types.OrderedMap[string, types.Type]) []wasm.ValueType {
	wasmParams := make([]wasm.ValueType, 0, params.Count())
	for _, paramType := range params.Iter() {
		wasmParams = append(wasmParams, typeToWASM(paramType))
	}
	return wasmParams
}

// Compile generates WASM module from the analyzed program and graph
func Compile(cfg Config) ([]byte, error) {
	ctx := core.NewContext(cfg.Analysis.Symbols, cfg.DisableHostImports)
	for _, item := range cfg.Program.AllTopLevelItem() {
		if err := compileTopLevelItem(ctx, item); err != nil {
			return nil, err
		}
	}
	return ctx.Module.Generate(), nil
}

// compileTopLevelItem dispatches compilation based on item type
func compileTopLevelItem(ctx *core.Context, item parser.ITopLevelItemContext) error {
	if funcDecl := item.FunctionDeclaration(); funcDecl != nil {
		return compileFunctionDeclaration(ctx, funcDecl)
	}
	if taskDecl := item.TaskDeclaration(); taskDecl != nil {
		return compileTaskDeclaration(ctx, taskDecl)
	}
	if flowStmt := item.FlowStatement(); flowStmt != nil {
		return nil
	}
	return errors.New("unknown top-level item")
}

// typeToWASM converts a Slate type to a WASM value type
func typeToWASM(t types.Type) wasm.ValueType {
	switch t.(type) {
	case types.I32:
		return wasm.I32
	case types.I64, types.TimeStamp, types.TimeSpan:
		return wasm.I64
	case types.F32:
		return wasm.F32
	case types.F64:
		return wasm.F64
	case types.U8, types.U16, types.U32, types.I8, types.I16:
		// All smaller integers are represented as i32 in WASM
		return wasm.I32
	case types.U64:
		return wasm.I64
	default:
		// Default to i32 for unknown types (bool, etc)
		return wasm.I32
	}
}

// compileFunctionDeclaration compiles a function definition
func compileFunctionDeclaration(
	ctx *core.Context,
	funcDecl parser.IFunctionDeclarationContext,
) error {
	name := funcDecl.IDENTIFIER().GetText()
	funcScope, err := ctx.Scope.Resolve(name)
	if err != nil {
		return errors.Wrapf(err, "function '%s' not found in symbols", name)
	}
	funcType, ok := funcScope.Type.(types.Function)
	if !ok {
		return errors.Newf("expected function type for '%s'", name)
	}

	params := make([]wasm.ValueType, 0, funcType.Params.Count())
	for _, paramType := range funcType.Params.Iter() {
		wasmType := typeToWASM(paramType)
		params = append(params, wasmType)
	}
	var results []wasm.ValueType
	if funcType.Return != nil {
		results = append(results, typeToWASM(funcType.Return))
	}
	typeIdx := ctx.Module.AddType(wasm.FunctionType{Params: params, Results: results})
	blockScope, _ := funcScope.FirstChildOfKind(symbol.KindBlock)
	blockCtx := ctx.WithScope(funcScope).WithNewWriter()
	if err = statement.CompileBlock(blockCtx, funcDecl.Block()); err != nil {
		return errors.Wrapf(err, "failed to compile function '%s' body", name)
	}
	funcIdx := ctx.Module.AddFunction(typeIdx, collectLocals(blockScope), blockCtx.Writer.Bytes())
	ctx.Module.AddExport(name, wasm.ExportFunc, funcIdx)
	return nil
}

func collectLocals(scope *symbol.Scope) []wasm.ValueType {
	var locals []wasm.ValueType
	for _, child := range scope.Children {
		if child.Symbol != nil && child.Kind == symbol.KindVariable {
			locals = append(locals, typeToWASM(child.Type))
		} else if child.Kind == symbol.KindBlock {
			locals = append(locals, collectLocals(child)...)
		}
	}
	return locals
}

// compileTaskDeclaration compiles a task definition
func compileTaskDeclaration(
	ctx *core.Context,
	taskDecl parser.ITaskDeclarationContext,
) error {
	name := taskDecl.IDENTIFIER().GetText()
	taskScope, err := ctx.Scope.Resolve(name)
	if err != nil {
		return errors.Wrapf(err, "task '%s' not found in symbols", name)
	}
	taskType, ok := taskScope.Type.(types.Task)
	if !ok {
		return errors.Newf("expected task type for '%s'", name)
	}
	wasmParams := append(paramsToWasm(taskType.Config), paramsToWasm(taskType.Params)...)
	var results []wasm.ValueType
	if taskType.Return != nil {
		results = append(results, typeToWASM(taskType.Return))
	}
	typeIdx := ctx.Module.AddType(wasm.FunctionType{Params: wasmParams, Results: results})
	blockScope, _ := taskScope.FirstChildOfKind(symbol.KindBlock)
	blockCtx := ctx.WithScope(blockScope).WithNewWriter()
	if err = statement.CompileBlock(blockCtx, taskDecl.Block()); err != nil {
		return errors.Wrapf(err, "failed to compile task '%s' body", name)
	}
	funcIdx := ctx.Module.AddFunction(typeIdx, collectLocals(blockScope), blockCtx.Writer.Bytes())
	ctx.Module.AddExport(name, wasm.ExportFunc, funcIdx)
	return nil
}
